src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,385 @@
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+  }
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Backward(
+    const MatType& input,
+    const MatType& gy,
+    MatType& g)
+{
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = inputDimension;
+
+  arma::Cube<typename MatType::elem_type> gyTemp(
+      const_cast<MatType&>(gy).memptr(), inputSize, size,
+      batchSize * higherDimension, false, false);
+  arma::Cube<typename MatType::elem_type> gTemp(
+      const_cast<MatType&>(g).memptr(), inputSize, size,
+      batchSize * higherDimension, false, false);
+
+  // Step 1: dl / dxhat.
+  arma::Cube<typename MatType::elem_type> norm =
+      gyTemp.each_slice() % arma::repmat(gamma.t(), inputSize, 1);
+
+  // Step 2: sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
+  MatType temp = arma::sum(norm % inputMean, 2);
+  MatType vars = temp % arma::repmat(arma::pow(stdInv, 3),
+      inputSize, 1) * -0.5;
+
+  // Step 3: dl / dxhat * 1 / stdInv + variance * 2 * (x - mu) / m +
+  // dl / dmu * 1 / m.
+  gTemp = (norm.each_slice() % arma::repmat(stdInv,
+      inputSize, 1) +
+      (inputMean.each_slice() % vars * 2)) / (batchSize * higherDimension);
+
+  // Step 4: sum (dl / dxhat * -1 / stdInv) + variance *
+  // (sum -2 * (x - mu)) / m.
+  MatType normTemp = arma::sum(norm.each_slice() %
+      arma::repmat(-stdInv, inputSize, 1) , 2) /
+      (batchSize * higherDimension);
+  gTemp.each_slice() += normTemp;
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Gradient(
+    const MatType& /* input */,
+    const MatType& error,
+    MatType& gradient)
+{
+  const size_t inputSize = inputDimension;
+
+  gradient.set_size(size + size, 1);
+  arma::Cube<typename MatType::elem_type> errorTemp(
+      const_cast<MatType&>(error).memptr(), inputSize, size,
+      error.n_cols, false, false);
+
+  // Step 5: dl / dy * xhat.
+  MatType temp = arma::sum(arma::sum(normalized % errorTemp, 0), 2);
+  gradient.submat(0, 0, gamma.n_elem - 1, 0) = temp.t();
+
+  // Step 6: dl / dy.
+  temp = arma::sum(arma::sum(errorTemp, 0), 2);
+  gradient.submat(gamma.n_elem, 0, gradient.n_elem - 1, 0) = temp.t();
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::ComputeOutputDimensions()
+{
+  if (minAxis > maxAxis)
+  {
+    Log::Fatal << "BatchNorm: minAxis must be less than or equal to maxAxis."
+        << std::endl;
+  }
+  this->outputDimensions = this->inputDimensions;
+  size_t mainMinAxis = std::min(this->inputDimensions.size() - 1, minAxis);
+  size_t mainMaxAxis = std::min(this->inputDimensions.size() - 1, maxAxis);
+  inputDimension = 1;
+  for (size_t i = 0; i < mainMinAxis; i++)
+    inputDimension *= this->inputDimensions[i];
+  size = this->inputDimensions[mainMinAxis];
+  for (size_t i = mainMinAxis + 1; i <= mainMaxAxis; i++)
+    size *= this->inputDimensions[i];
```

Comment:
```suggestion

  size = this->inputDimensions[mainMinAxis];
  for (size_t i = mainMinAxis + 1; i <= mainMaxAxis; i++)
    size *= this->inputDimensions[i];

```
I think a little space here can help the readability (and feel free to add a comment too), but up to you if you want to take the suggestion. :+1: